### Тестовое задание на стажировку в ВК

Подготовка и проверка датасета.

In [1]:
import pandas as pd
import numpy as np
import sweetviz as sv

from catboost import CatBoostRanker, Pool, MetricVisualizer
import xgboost as xgb
from sklearn.metrics import ndcg_score
from sklearn.metrics import top_k_accuracy_score

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


In [55]:
df = pd.read_csv('data/intern_task.csv')

In [3]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None) 

In [4]:
# pd.reset_option('display.max_rows')
# pd.reset_option('display.max_columns') 

In [5]:
df.head(100)

rank  query_id  feature_0  feature_1  feature_2  feature_3  feature_4  \
0      0        10        1.0        0.0        1.0        3.0        3.0   
1      1        10        3.0        0.0        3.0        0.0        3.0   
2      0        10        3.0        0.0        2.0        0.0        3.0   
3      1        10        3.0        0.0        3.0        0.0        3.0   
4      2        10        3.0        0.0        3.0        1.0        3.0   
5      1        10        3.0        0.0        3.0        1.0        3.0   
6      1        10        3.0        0.0        3.0        0.0        3.0   
7      0        10        0.0        0.0        0.0        0.0        0.0   
8      0        10        2.0        0.0        1.0        0.0        2.0   
9      0        10        3.0        0.0        0.0        0.0        3.0   
10     1        10        3.0        1.0        3.0        0.0        3.0   
11     0        10        3.0        0.0        3.0        1.0        3.0   
12     1        10        3.0        0.0        3.0        0.0        3.0   
13     0        10        3.0        0.0        3.0        1.0        3.0   
14     0        10        3.0        0.0        3.0        0.0        3.0   
15     0        10        3.0        3.0        3.0        0.0        3.0   
16     1        10        3.0        0.0        3.0        1.0        3.0   
17     0        10        1.0        0.0        0.0        0.0        1.0   
18     1        10        3.0        3.0        3.0        0.0        3.0   
19     1        10        3.0        0.0        3.0        0.0        3.0   
20     1        10        3.0        0.0        3.0        0.0        3.0   
21     2        10        3.0        0.0        3.0        0.0        3.0   
22     4        10        3.0        0.0        3.0        0.0        3.0   
23     0        10        3.0        0.0        3.0        0.0        3.0   
24     4        10        3.0        0.0        3.0        0.0        3.0   
25     0        10        0.0        0.0        0.0        0.0        0.0   
26     0        10        2.0        0.0        2.0        0.0        2.0   
27     0        10        0.0        0.0        0.0        0.0        0.0   
28     1        10        3.0        0.0        0.0        0.0        3.0   
29     0        10        3.0        0.0        3.0        1.0        3.0   
30     0        10        3.0        0.0        3.0        1.0        3.0   
31     0        10        3.0        0.0        3.0        0.0        3.0   
32     0        10        0.0        0.0        0.0        0.0        0.0   
33     0        10        3.0        0.0        3.0        0.0        3.0   
34     1        10        3.0        0.0        3.0        0.0        3.0   
35     1        10        3.0        0.0        1.0        1.0        3.0   
36     0        10        0.0        0.0        0.0        0.0        0.0   
37     1        10        3.0        0.0        3.0        1.0        3.0   
38     0        10        1.0        0.0        1.0        0.0        1.0   
39     0        10        1.0        0.0        0.0        0.0        1.0   
40     2        10        3.0        0.0        3.0        0.0        3.0   
41     0        10        3.0        0.0        2.0        0.0        3.0   
42     0        10        3.0        0.0        1.0        1.0        3.0   
43     1        10        3.0        0.0        3.0        0.0        3.0   
44     2        10        3.0        0.0        1.0        0.0        3.0   
45     1        10        3.0        0.0        3.0        0.0        3.0   
46     1        10        3.0        0.0        0.0        1.0        3.0   
47     2        10        3.0        0.0        3.0        1.0        3.0   
48     0        10        3.0        0.0        3.0        1.0        3.0   
49     1        10        3.0        0.0        0.0        0.0        3.0   
50     0        10        3.0        0.0        0.0        0.0        3.0   


In [6]:
df.shape

(235258, 146)

In [7]:
df.describe().T

count          mean           std          min            25%  \
rank         235258.0  6.778686e-01  8.300334e-01     0.000000       0.000000   
query_id     235258.0  1.482841e+04  8.193945e+03    10.000000    8215.000000   
feature_0    235258.0  1.911960e+00  1.237374e+00     0.000000       1.000000   
feature_1    235258.0  2.062332e-01  5.790888e-01     0.000000       0.000000   
feature_2    235258.0  1.189847e+00  1.037233e+00     0.000000       0.000000   
feature_3    235258.0  5.502725e-01  7.909465e-01     0.000000       0.000000   
feature_4    235258.0  1.960082e+00  1.203534e+00     0.000000       1.000000   
feature_5    235258.0  8.039641e-01  3.399551e-01     0.000000       0.666667   
feature_6    235258.0  9.755747e-02  2.660348e-01     0.000000       0.000000   
feature_7    235258.0  5.411739e-01  4.199727e-01     0.000000       0.000000   
feature_8    235258.0  5.000000e+02  2.886770e+02     0.000000     250.000000   
feature_9    235258.0  8.349089e-01  3.107624e-01     0.000000       0.750000   
feature_10   235258.0  6.635420e+02  7.814060e+02     0.000000     207.000000   
feature_11   235258.0  1.526142e+00  1.541642e+01     0.000000       0.000000   
feature_12   235258.0  8.411803e+00  5.894885e+01     0.000000       4.000000   
feature_13   235258.0  7.921308e+00  3.449127e+00     2.000000       6.000000   
feature_14   235258.0  6.814013e+02  7.883018e+02     2.000000     222.000000   
feature_15   235258.0  1.028386e+01  6.101613e+00    -4.387408       6.191777   
feature_16   235258.0  2.068822e+01  9.661968e+00     5.154618      14.528860   
feature_17   235258.0  1.864941e+01  8.932978e+00     4.361284      12.722823   
feature_18   235258.0  1.999401e+01  9.876109e+00     4.018781      13.717640   
feature_19   235258.0  1.026942e+01  6.096088e+00    -4.412168       6.164914   
feature_20   235258.0  5.000000e-01  2.886770e-01     0.000000       0.250000   
feature_21   235258.0  3.189775e-01  1.424956e+00     0.000000       0.000000   
feature_22   235258.0  1.503966e+00  2.135136e+00     0.000000       0.000000   
feature_23   235258.0  5.910235e-01  8.905115e-01     0.000000       0.000000   
feature_24   235258.0  2.487188e+01  6.457217e+01     0.000000       5.000000   
feature_25   235258.0  5.392675e+00  1.134243e+01     0.000000       0.000000   
feature_26   235258.0  9.249845e-02  5.030169e-01     0.000000       0.000000   
feature_27   235258.0  4.451538e-01  7.495146e-01     0.000000       0.000000   
feature_28   235258.0  1.935110e-01  4.266477e-01     0.000000       0.000000   
feature_29   235258.0  6.194387e+00  1.194383e+01     0.000000       0.000000   
feature_30   235258.0  1.464715e+01  2.364352e+01     0.000000       3.000000   
feature_31   235258.0  2.204473e-01  8.465625e-01     0.000000       0.000000   
feature_32   235258.0  9.387608e-01  1.389356e+00     0.000000       0.000000   
feature_33   235258.0  4.333073e-01  5.753261e-01     0.000000       0.000000   
feature_34   235258.0  1.613138e+01  2.419105e+01     0.000000       4.000000   
feature_35   235258.0  1.000000e+06  5.773540e+05     0.000000  500000.000000   
feature_36   235258.0  1.523726e-01  6.111870e-01     0.000000       0.000000   
feature_37   235258.0  6.808744e-01  8.521698e-01     0.000000       0.000000   
feature_38   235258.0  3.017343e-01  4.458368e-01     0.000000       0.000000   
feature_39   235258.0  1.069536e+01  1.483763e+01     0.000000       2.750000   
feature_40   235258.0  1.015913e+02  1.560292e+03     0.000000       0.000000   
feature_41   235258.0  9.376932e-02  2.903695e+00     0.000000       0.000000   
feature_42   235258.0  3.729501e-01  2.652779e+01     0.000000       0.000000   
feature_43   235258.0  6.277281e-02  1.596514e-01     0.000000       0.000000   
feature_44   235258.0  1.087270e+02  1.585344e+03     0.000000       0.000000   
feature_45   235258.0  5.059999e-02  7.001287e-02     0.000000       0.010819   
feature_46   235258.0  6.845343e-02  2.042248e-

С целью ознакомления с датасетом проведем разведочный анализ.  
Для это воспользуемся готовой библиотекой sweetviz.

In [8]:
report = sv.analyze([df, "Permit Logs"])

                                             |                  | [  0%]   00:00 -> (? left)

C:\ProgramData\Anaconda3\lib\site-packages\sweetviz\graph_associations.py:219: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  new_dataframe[feature] = pd.Series(dtype=float)
C:\ProgramData\Anaconda3\lib\site-packages\sweetviz\graph_associations.py:219: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  new_dataframe[feature] = pd.Series(dtype=float)
C:\ProgramData\Anaconda3\lib\site-packages\sweetviz\graph_associations.py:219: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` man

In [9]:
report.show_html('report.html')

Report report.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


Полученные отчеты находятся в папке eda_reports.

In [10]:
from IPython.display import display, HTML, IFrame

In [11]:
IFrame(src = 'report.html', width = 1000, height = 500)

При изучении отчетов EDA, было выявлено 13 категориальных признаков, включая константы  
(feature_64, feature_65, feature_72, feature_100),  
бинарные (feature_95, feature_96, feature_97, feature_98, feature_99).  
Остальные признаки, определенные как категориальные, имеют до 10 значений.

In [64]:
const_cols_df = df.loc[:, df.nunique() == 1].columns.to_list()
const_cols_df

['feature_64', 'feature_65', 'feature_72', 'feature_100']

Удалим категориальные признаки

In [65]:
df.drop(const_cols_df, axis=1, inplace=True)

Составим матрицу корреляции Пирсона и рассмотрип признаки с наибольшей корреляцией

In [14]:
# df.corr()

In [15]:
def correl(X_train, thresh):
    cor = X_train.corr()
    c1 = cor.stack().sort_values(ascending=False).drop_duplicates()
    all_cor = c1[c1.values != 1]
    return all_cor[abs(all_cor) > thresh]

In [16]:
correl(df, 0.95)

feature_35   feature_8      1.000000
feature_20   feature_8      1.000000
feature_35   feature_20     1.000000
feature_19   feature_15     0.999892
feature_74   feature_70     0.999599
feature_90   feature_94     0.999323
feature_20   query_id       0.999009
feature_35   query_id       0.999009
query_id     feature_8      0.999009
feature_24   feature_142    0.999007
feature_123  feature_113    0.997952
feature_111  feature_121    0.997734
feature_119  feature_115    0.997582
feature_44   feature_40     0.997458
feature_34   feature_30     0.997357
feature_80   feature_84     0.997088
feature_10   feature_14     0.997014
feature_39   feature_143    0.996818
feature_85   feature_89     0.996235
feature_114  feature_124    0.996158
feature_29   feature_25     0.996002
feature_75   feature_79     0.995725
feature_120  feature_110    0.995705
feature_122  feature_112    0.994764
feature_16   feature_17     0.991783
feature_110  feature_114    0.991119
feature_120  feature_124    0.989141
f

Разделим датсет на тренировочную и тестовую выборку так, чтобы в выборку попали объекты с уникальным query_id.

In [66]:
df['query_id'].nunique()

2000

In [67]:
grouped = df.groupby('query_id').sum().reset_index()

In [68]:
grouped.iloc[1499,0], grouped.iloc[1500,0]

(22495, 22510)

In [69]:
df_train = df[df['query_id'] <= 22495]
df_test = df[df['query_id'] > 22495]

In [70]:
X_train = df_train.drop(["rank", "query_id"], axis=1)
y_train = df_train.iloc[:,0]
queries_train = df_train.iloc[:,1]

In [71]:
X_test = df_test.drop(["rank"], axis=1)
y_test = df_test.iloc[:,0]
queries_test = df_test.iloc[:,1]

Для расчета такие показатели, как NDCG и найденная релевантность, должны находиться в сегменте [0,1].

Статистика помечена значениями значимости

0 - неактуально, 1 - очень важно. В таблице представлено количество документов для каждого значения.

In [72]:
from collections import Counter
Counter(y_train).items()

dict_items([(0, 100588), (1, 57486), (2, 23561), (4, 1215), (3, 2938)])

In [73]:
max_relevance = np.max(y_train)
y_train = y_train / max_relevance
y_test = y_test / max_relevance

Создаим простую модель CatBoost

In [27]:
train = Pool(
    data=X_train,
    label=y_train,
    group_id=queries_train
)

test = Pool(
    data=X_test,
    label=y_test,
    group_id=queries_test
)

In [28]:
model_cb = CatBoostRanker(
    metric_period = 20,
    max_depth =  10,
    eval_metric = 'NDCG:top=5',
    random_seed = 42,
    task_type = "GPU"
)
model_cb.fit(
    train,
    eval_set = test,
    verbose=False
)

Metric PFound is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=5;type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


In [29]:
# NDCG@5 на тренировочной выборке

y_pred_train = model_cb.predict(X_train)
ndcg_score([y_train], [y_pred_train], k=5)

0.9999999999999999

In [30]:
# NDCG@5 на тестовой выборке

y_pred_test = model_cb.predict(X_test)
ndcg_score([y_test], [y_pred_test], k=5)

0.6854018932919382

Создаим простую модель XGBoost

In [34]:
import xgboost as xgb

xgb.__version__

'2.0.3'

In [79]:
df_train = df[df['query_id'] <= 22495]
df_test = df[df['query_id'] > 22495]

In [80]:
X_train = df_train.drop(columns=['rank', 'query_id'])
y_train = df_train['rank']
queries_train = df_train['query_id']

In [81]:
X_test = df_test.drop(columns=['rank', 'query_id'])
y_test = df_test['rank']
queries_test = df_test['query_id']


In [82]:
model_xgb = xgb.XGBRanker(
    max_depth =  10,
    objective = 'rank:ndcg',
    eval_metric=["ndcg"],
    seed = 42,
    device='gpu'
)
model_xgb.fit(
    X_train,
    y_train,
    qid=queries_train,
    verbose=False
)

XGBRanker(base_score=None, booster=None, callbacks=None, colsample_bylevel=None,
          colsample_bynode=None, colsample_bytree=None, device='gpu',
          early_stopping_rounds=None, enable_categorical=False,
          eval_metric=['ndcg'], feature_types=None, gamma=None,
          grow_policy=None, importance_type=None, interaction_constraints=None,
          learning_rate=None, max_bin=None, max_cat_threshold=None,
          max_cat_to_onehot=None, max_delta_step=None, max_depth=10,
          max_leaves=None, min_child_weight=None, missing=nan,
          monotone_constraints=None, multi_strategy=None, n_estimators=None,
          n_jobs=None, num_parallel_tree=None, random_state=None, ...)

In [85]:
# NDCG@5 на тренировочной выборке

y_pred_tain = model_xgb.predict(X_train)
ndcg_score([y_train], [y_pred_tain], k=5)

C:\ProgramData\Anaconda3\lib\site-packages\xgboost\core.py:160: UserWarning: [23:09:09] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0b3782d1791676daf-1\xgboost\xgboost-ci-windows\src\common\error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.
Potential solutions:
- Use a data structure that matches the device ordinal in the booster.
- Set the device for booster before call to inplace_predict.

This warning will only be shown once.

  warnings.warn(smsg, UserWarning)


0.9999999999999999

In [86]:
# NDCG@5 на тестовой выборке

y_pred_test = model_xgb.predict(X_test)
ndcg_score([y_test], [y_pred_test])

0.9268140700396167

В этой работе я сравнил результаты библиотек CatBoost, XGBoost.  
Метрики XGBoost получились лучше, чем у CatBoost.  
Предположительно это связано с переобучением модели.  

|     | CatBoost | XGB |
| --- | --- | --- |
| train | 0.99 | 0.69 |
|test | 0.99 | 0.93 |


В качестве улучшений проделанной работы можно рассмотреть работу в следующих направлениях:

- отбор признаков датасета,
- генерация новых признаков,
- собрать ансамбль ранжирующих моделей.